docs/end-to-end-demo.md

```diff
@@ -1,8 +1,36 @@
 # End-to-End Demo
 In this guide, you will deploy Akri end-to-end, all the way from discovering local video cameras to the footage being streamed on a Web application. You will explore how Akri can dynamically discover devices, deploy brokers pods to perform some action on a device (in this case grabbing video frames and serving them over gRPC), and deploy broker services for obtaining the results of that action.
 
-**Note:** The first step can be performed either via K3s or MicroK8s. Select and
-carry out one or the other, then continue on with the rest of the steps. 
+## Set up mock udev video devices
+1. Install a kernel module to make v4l2 loopback video devices. Learn more about this module [here](https://github.com/umlaeute/v4l2loopback).
+    ```sh
+    curl http://deb.debian.org/debian/pool/main/v/v4l2loopback/v4l2loopback-dkms_0.12.5-1_all.deb -o v4l2loopback-dkms_0.12.5-1_all.deb
+    sudo dpkg -i v4l2loopback-dkms_0.12.5-1_all.deb
+    ```
+1. Insert the kernel module, creating /dev/video1 and /dev/video2 devnodes. To create different number video devices modify the `video_nr` argument. 
+    ```sh
+    sudo modprobe v4l2loopback exclusive_caps=1 video_nr=1,2
+    ```
+1. Install Gstreamer main packages
+    ```sh
+    sudo apt-get install -y \
+        libgstreamer1.0-0 gstreamer1.0-tools gstreamer1.0-plugins-base \
+        gstreamer1.0-plugins-good gstreamer1.0-libav
+    ```
+1. Open two new terminals (one for each fake video device), and in each terminal ssh into your ubuntu server that your cluster is running on.
+1. In one terminal, stream a test video of a white ball moving around a black background from the first fake video device.
+    ```sh
+    sudo gst-launch-1.0 -v videotestsrc pattern=ball ! "video/x-raw,width=640,height=480,framerate=10/1" ! avenc_mjpeg ! v4l2sink device=/dev/video1
+    ```
+    If this generates an error, be sure that there are no existing video streams targeting /dev/video1 (you can query with commands like this: `ps -aux | grep gst-launch-1.0 | grep "/dev/video1"`).
+1. In the other terminal, stream a test video of SMPTE 100%% color bars moving horizontally from the second fake video device.
+    ```sh
+    sudo gst-launch-1.0 -v videotestsrc pattern=smpte horizontal-speed=1 ! "video/x-raw,width=640,height=480,framerate=10/1" ! avenc_mjpeg ! v4l2sink device=/dev/video2
+    ```
+    If this generates an error, be sure that there are no existing video streams targeting /dev/video2 (you can query with commands like this: `ps -aux | grep gst-launch-1.0 | grep "/dev/video2"`).
+
+**Note:** Feel free to deploy on any Kubernetes distribution. Here, find instructions for K3s and MicroK8s. Select and
```

Comment:
i think this could be kind of hidden. Can we rearrange this so that the header is ## Set up a cluster and the subheaders are the two options?

Reply:
You got it, coming up shortly.